### PR TITLE
[mlir] Improve `GreedyPatternRewriteDriver` and pass documentation

### PR DESCRIPTION
**Pull request by matthias-springer**
Clarify what kind of IR modifications are allowed. Also improve the documentation of the greedy rewrite driver entry points.

Addressing comments in #76219.

### CODE REVIEW

File: mlir/docs/PassManagement.md [hunk 1]
```diff
@@ -23,16 +23,21 @@ further below. All passes in MLIR derive from `OperationPass` and adhere to the
...
+    siblings.
     *   Other threads may be modifying these operations in parallel.
     *   Inspecting the state of ancestor/parent operations is permitted.
+*   Must not modify the state of operations other than the operation that the
+    pass operates on ("current operation") and its nested operations. This
+    includes adding, modifying or removing other operations from an ancestor
+    block.
+    *   Other threads may be operating on these operations simultaneously.
+    *   The attributes of the current operation may be modified freely.
+    *   The operands of the current operation may be modified, as long as no
```
Comment (matthias-springer) [inline mlir/docs/PassManagement.md:37]: This is a bit sus. Example:

- Pass instance 1: `currentOp->setOperand(0, currentOp->getBlock()->getArgument(0)`
- Pass instance 2: `print(currentOp->getBlock()->getArgument(0).getUses().size()`

Instance 1 modifies the current op. That is allowed. (Or is it? The use list of a parent block argument changes. Does this count as "modifying per parent op"?).

Instance 2 reads the number of uses of a parent block argument. That is allowed. But there's a race condition if both instances run in parallel.

Maybe the only modifications that we should allow are "modifying attributes"?

Comment (joker-eph) [reply mlir/docs/PassManagement.md:37]: You can't touch operands, this would race in the parent block.

Comment (matthias-springer) [reply mlir/docs/PassManagement.md]: Ok, that's what I thought... So the only permissible change to the operation that the pass operates on is "modifying attributes". I updates the docs accordingly.

**Review by joker-eph — APPROVED** [review]

**Comment by jpienaar** [comment]
Thanks!

**Pull request merged**